### DISCUSSION TITLE
[Question] Recursive Reference of Entities to themselves with TS 4.0+

### DISCUSSION THREAD

**Discussion by darkbasic** [mikro-orm/mikro-orm#2413] [Q&A]
This issue is not specific to mikro-orm, but prevents from creating Mikro-ORM adapters for third-party libraries so I think this may be a good place to seek for help.

This is the adapter I'm building: https://github.com/darkbasic/accounts/tree/mikro-orm-v1/packages/database-mikro-orm
And this is the example which makes use of it: https://github.com/darkbasic/accounts/tree/mikro-orm-v1/examples/graphql-server-mikro-orm-postgres

(the README mentions yarn but it uses pnpm, it's an old PR and I still need to update it)

In the adapter I'm basically creating some base entities (User, Email, Service, Session) with several relationships between them. These entities are supposed to be extended and customized by the user, the library should just provide the bare minimum for working authentication.
Extending them is exactly what I am doing in the example and everything used to work flawlessly up uo Typescript 3.9, but 4.0+ regressed:  https://github.com/microsoft/TypeScript/issues/40315

Now I'm getting this errors:
```
$ pnpm tsc
src/entities/email.ts:13:14 - error TS2310: Type 'Email' recursively references itself as a base type.

13 export class Email extends AccountsEmail {
                ~~~~~

src/entities/email.ts:13:14 - error TS2506: 'Email' is referenced directly or indirectly in its own base expression.

13 export class Email extends AccountsEmail {
                ~~~~~

src/entities/user.ts:22:3 - error TS2344: Type 'Email' does not satisfy the constraint 'Email<any>'.
  Type 'Email' is missing the following properties from type 'Email<any>': id, user, address, verified

22   Email,
     ~~~~~

src/entities/user.ts:27:3 - error TS2344: Type 'Email' does not satisfy the constraint 'Email<any>'.
  Type 'Email' is missing the following properties from type 'Email<any>': id, user, address, verified

27   Email,
     ~~~~~

src/index.ts:24:67 - error TS2419: Types of construct signatures are incompatible.
  Type 'new ({ fullName, ...otherProps }: EmailCtorArgs) => Email' is not assignable to type 'new (...args: any[]) => Email<any>'.
    Type 'Email' is not assignable to type 'Email<any>'.

24   const accountsDb = new AccountsMikroOrm({ em, UserEntity: User, EmailEntity: Email });
                                                                     ~~~~~~~~~~~

  ../../packages/database-mikro-orm/lib/types/index.d.ts:9:5
    9     EmailEntity?: Constructor<CustomEmail | Email<any>>;
          ~~~~~~~~~~~
    The expected type comes from property 'EmailEntity' which is declared here on type 'AccountsMikroOrmOptions<User, Email<any>, Session<any>, Service<any>>'

src/mikro-orm-config.ts:18:35 - error TS2419: Types of construct signatures are incompatible.
  Type 'new ({ fullName, ...otherProps }: EmailCtorArgs) => Email' is not assignable to type 'new (args: EmailCtorArgs<any>) => Email<any>'.
    Type 'Email' is not assignable to type 'Email<any>'.

18     getUserSchema({ AccountsUser, EmailEntity: Email, abstract: true }),
                                     ~~~~~~~~~~~

  ../../packages/database-mikro-orm/lib/entity/User.d.ts:30:5
    30     EmailEntity?: EmailCtor<any> | undefined;
           ~~~~~~~~~~~
    The expected type comes from property 'EmailEntity' which is declared here on type '{ AccountsUser: UserCtor<any, any, any, any>; EmailEntity?: EmailCtor<any>; SessionEntity?: SessionCtor<any>; ServiceEntity?: ServiceCtor<any>; abstract?: boolean; }'


Found 6 errors.
```

Reproducing the issue is as easy as:
```
git clone https://github.com/darkbasic/accounts.git
cd accounts
pnpm install && pnpm compile
cd examples/graphql-server-mikro-orm-postgres
pnpm tsc
```

I really don't know how to fix it and it would be such a pity to discard all this work because of this, because I've tried to get a Mikro-ORM adapter merged for more than a year and because of issues with other libraries I've never been able to. Now that those are finally fixed Typescript is ruining the day...

**Comment by B4nan** [comment]
I'd say you should try to strip down the ORM parts and make this pure TS issue - then seek help in the TS issues/discord. Right now it's a bit hard to wrap your head around and I have no time to deep dive into this, the priority is to get v5 stable.

Will keep this open for few days before moving to discussions, but I don't think this is actionable from the ORM point of view 🤷

> I really don't know how to fix it and it would be such a pity to discard all this work because of this, because I've tried to get a Mikro-ORM adapter merged for more than a year and because of issues with other libraries I've never been able to. Now that those are finally fixed Typescript is ruining the day...

It's TS issue, right? You could always make the types less strict/generic, to get around it, right?

**Comment by darkbasic** [comment]
> You could always make the types less strict/generic, to get around it, right?

Not without losing access to all the Mikro-ORM TS helpers: I need relationships between entities.
`import type` helped alot to alleviate the issue, but there are still points where I can't use it: https://github.com/darkbasic/accounts/blob/mikro-orm-v1/examples/graphql-server-mikro-orm-postgres/src/entities/user.ts#L35
The code which gets called: https://github.com/darkbasic/accounts/blob/mikro-orm-v1/packages/database-mikro-orm/src/entity/User.ts#L30

As you can see I have been able to work it around with `import type`s for almost everything expect this function, which still needs access to the value (and thus needs a traditional import).

The same is also true for `getUserSchema`: https://github.com/darkbasic/accounts/blob/mikro-orm-v1/examples/graphql-server-mikro-orm-postgres/src/mikro-orm-config.ts#L18

> I'd say you should try to strip down the ORM parts and make this pure TS issue - then seek help in the TS issues/discord.

I'll try.
